pkg/oc/admin/diagnostics/diagnostics.go

```diff
@@ -82,50 +65,100 @@ const (
 
 		    %[1]s --master-config=/etc/origin/master/master-config.yaml
 
-		* If master/node config files are not found and the --host flag is not
-		  present, host diagnostics are skipped.
-		* If the client has cluster-admin access, this access enables cluster
-		  diagnostics to run which regular users cannot.
-		* If a client config file is not found, client and cluster diagnostics
-		  are skipped.
+		* Explicitly specifying a config file raises an error if it is not found.
+		* A client config with cluster-admin access is required for most cluster diagnostics.
+		* Diagnostics that require a config file are skipped if it is not found.
+		* The standard set also skips diagnostics considered too heavyweight.
 
-		Diagnostics may be individually run by passing diagnostic name as arguments.
+		An individual diagnostic may be run as a subcommand which may have flags
+		for specifying options specific to that diagnostic.
 
-		    %[1]s <DiagnosticName>
+		Finally, the "all" subcommand runs all available diagnostics (including heavyweight
+		ones skipped in the standard set) and provides all individual diagnostic flags.
+		`)
+	longDescriptionAll = templates.LongDesc(`
+		This utility helps troubleshoot and diagnose known problems for an OpenShift cluster
+		and/or local host. This subcommand exists to run all available diagnostics:
 
-		The available diagnostic names are: %[2]s.`)
+		    %[1]s
+
+		Available diagnostics vary based on client config and local OpenShift host config.
+		All flags from the base command work similarly here, but all possible flags for
+		individual diagnostics are also available.
+		`)
+	longDescriptionIndividual = templates.LongDesc(`
+		Runs the %s diagnostic.
+
+		%s
+		`)
 )
 
 // NewCmdDiagnostics is the base command for running any diagnostics.
 func NewCmdDiagnostics(name string, fullName string, out io.Writer) *cobra.Command {
+	available := availableDiagnostics()
 	o := &DiagnosticsOptions{
-		RequestedDiagnostics: []string{},
-		LogOptions:           &log.LoggerOptions{Out: out},
-		ImageTemplate:        variable.NewDefaultImageTemplate(),
-		NetworkOptions:       &NetworkDiagnosticsOptions{},
+		RequestedDiagnostics:     available.Names().Difference(defaultSkipDiagnostics()),
+		ParameterizedDiagnostics: types.NewParameterizedDiagnosticMap(available...),
+		LogOptions:               &log.LoggerOptions{Out: out},
 	}
 
 	cmd := &cobra.Command{
 		Use:   name,
 		Short: "Diagnose common cluster problems",
-		Long:  fmt.Sprintf(longDescription, fullName, strings.Join(availableDiagnostics().List(), ", ")),
+		Long:  fmt.Sprintf(longDescription, fullName),
 		Run: func(c *cobra.Command, args []string) {
-			kcmdutil.CheckErr(o.Complete(args))
+			kcmdutil.CheckErr(o.Complete(c, args))
+
+			failed, err, warnCount, errorCount := o.RunDiagnostics()
+			o.Logger.Summary(warnCount, errorCount)
+
+			kcmdutil.CheckErr(err)
+			if failed {
+				os.Exit(255)
```

Comment:
is this exit code documented somewhere by any chance?

Reply:
No. Diagnostics exit code only indicates success/failure with no further nuance. Is there a better code to use?

Reply:
Hm, could not really find any relevant docs; only this from v2: https://access.redhat.com/documentation/en-US/OpenShift_Online/2.0/html/Cartridge_Specification_Guide/Exit_Status_Codes.html

We do use the `255` exit code elsewhere in our code, but it only seems to be when starting a node, master, etc. from a set of given invalid options [1].

Would an exit code of `1` make more sense? (Although I would not see this as a blocking change)

1. https://github.com/openshift/origin/blob/master/pkg/cmd/server/start/start_allinone.go#L96

Reply:
It makes no difference to me. I'll change it to 1 as I'm making other tweaks anyway.